### PR TITLE
CD-240: Clearone Script

### PR DESCRIPTION
**Pull request by mariemm97**

### CODE REVIEW

File: examples/http/clearone_convergence.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            console.error(error)
+            D.failure(D.errorType.GENERIC_ERROR);
+        }
+        d.resolve(JSON.parse(body));
+    });
+    return d.promise;
+}
+
+// Get device information
+function getData() {
```
Comment (mohamedkallel) [inline examples/http/clearone_convergence.js:67]: for a clearer code extract split this function to 2 separate promises, the first for 
httpGet("/dashboard-back/api/devices/pro-audio/search?q=" + D.device.ip())
if this is successfull call the next 
httpGet("/dashboard-back/api/devices/pro-audio/" + data.avDevices[0].id)

if we name the first one search() and the second one getDetail() the get status will be like this

search().
    then(getDetail).
    then(extractVariables)...

**Review by mohamedkallel — CHANGES_REQUESTED** [review]

File: examples/http/clearone_convergence.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+/**
+ * @param {string} url  The URL to perform the GET request
+ * @returns promise for http response body
+ */
+function httpGet(url) {
+    var d = D.q.defer();
+	var website = D.createExternalDevice(urlCl);
+    var config = {
+        url: url,
+		protocol: "http:",
```
Comment (GeorgiTsatsev) [inline examples/http/clearone_convergence.js:41]: The default protocol is http, this should not be needed

Comment (GeorgiTsatsev) [inline examples/http/clearone_convergence.js:41]: Formatting appears wrong

File: examples/http/clearone_convergence.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    };
+    website.http.get(config, function (error, response, body) {
+		if (error) {
+            console.error(error)
+            D.failure(D.errorType.GENERIC_ERROR);
+        }
+        if (response.statusCode == 404) {
+            D.failure(D.errorType.RESOURCE_UNAVAILABLE);
+        }
+        if (response.statusCode === 401 || response.statusCode === 403) {
```
Comment (GeorgiTsatsev) [inline examples/http/clearone_convergence.js:55]: status code checks should be else ifs

File: examples/http/clearone_convergence.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    return d.promise;
+}
+
+/**
+ * @returns promise for http response body containig device information
+ */
+function search() {
+    return httpGet("/dashboard-back/api/devices/pro-audio/search?q=" + D.device.ip())
+        .then(function(data){
+            deviceInfo = data.avDevices[0];
```
Comment (GeorgiTsatsev) [inline examples/http/clearone_convergence.js:73]: what if the avDevices is not part of the data response? (eg an error or no avDevices in the response body at all)

File: examples/http/clearone_convergence.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        });
+}
+
+// Function to extract variables
+function extractVariables() {
+    var firmware = deviceInfo.firmwareVersion;
+    if (deviceInfo.hasNewFirmware) {
+        firmware = firmware + ' *';
+    }
+    var variables = [
```
Comment (GeorgiTsatsev) [inline examples/http/clearone_convergence.js:93]: maybe add typing

File: examples/http/clearone_convergence.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+/**
+ * @remote_procedure
+ * @label Validate ClearOne Device
+ * @documentation This procedure is used to validate if data is accessible
+ */
+function validate() {
+    search()
+        .then(D.success)
```
Comment (GeorgiTsatsev) [inline examples/http/clearone_convergence.js:112]: the search does not check for any errors in the response data

**Review by GeorgiTsatsev — APPROVED** [review]

**Review by GeorgiTsatsev — APPROVED** [review]

**Review by mohamedkallel — APPROVED** [review]

**Review by GeorgiTsatsev — APPROVED** [review]

**Pull request merged**